### ISSUE TITLE
Request Header Fields Too Large

### ISSUE THREAD

**Issue by jengle-dev** [jengle-dev/JJR-Bookstore#37]
![image](https://github.com/jengle-dev/Bookstore/assets/117794203/d5c57235-99cb-43ff-b02b-438087eac2de)

**Comment by jengle-dev** [comment]
ChatGPT Response to asking about 431 response: 

The HTTP status code 431 indicates that the server is refusing to process the request because the request headers are too large. This error typically occurs when the server or backend you are communicating with has a limitation on the maximum size of request headers it can handle.

To address this issue, you can try the following steps:

1. Review the request headers being sent from your client application. Check if there are any headers that contain large amounts of data or unnecessary headers that can be removed. Consider reducing the size of the headers by removing any unnecessary or redundant information.

2. If you are using a library or framework to send the requests, such as Axios or the Fetch API, check if there are any options or configurations related to request headers. Look for options that allow you to modify or limit the size of the headers being sent.

3. If you have control over the server or backend application, you can increase the maximum size limit for request headers. Consult the documentation or configuration files of the server framework or web server software you are using to adjust the maximum header size.

4. If you are using a reverse proxy or load balancer in front of your server, such as Nginx or Apache, check their configurations. There might be a limitation set on the maximum size of request headers that needs to be adjusted.

5. Consider compressing the request headers if possible. Some servers and frameworks support header compression, which can reduce the overall size of the headers being sent.

6. If the issue persists, consult the documentation or support resources of the server or backend application you are working with. They might provide specific recommendations or workarounds for handling large request headers.

By reviewing and implementing these steps, you should be able to address the "431 Request Header Fields Too Large" error and allow the server to process your requests successfully.

**Comment by jengle-dev** [comment]
Another ChatGPT response: 

In the provided code snippet, there doesn't appear to be a specific section for setting or manipulating request headers. However, request headers are typically managed when making HTTP requests using libraries like `fetch`, Axios, or Apollo Client.

In the code you provided, you are using Apollo Client to make GraphQL requests, and Apollo Client automatically manages the necessary headers for sending GraphQL queries and mutations.

If you need to modify or add custom headers to your requests, you can do so by configuring Apollo Client's `HttpLink` with a `headers` property. Here's an example of how you can add custom headers to Apollo Client:

```jsx
import { ApolloClient, ApolloProvider, InMemoryCache, createHttpLink } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';

const httpLink = createHttpLink({
  uri: '/graphql',
});

const authLink = setContext((_, { headers }) => {
  // Add your custom headers here
  const customHeaders = {
    'Authorization': 'Bearer YOUR_TOKEN',
    'Custom-Header': 'Custom Value',
  };

  return {
    headers: {
      ...headers,
      ...customHeaders,
    },
  };
});

const client = new ApolloClient({
  link: authLink.concat(httpLink),
  cache: new InMemoryCache(),
});

// Rest of your code...
```

In this example, the `setContext` function from `@apollo/client/link/context` is used to add custom headers to each request. You can modify the `customHeaders` object to include the specific headers you want to send. Make sure to adjust the values to match your desired header names and values.

By configuring Apollo Client with the appropriate `headers`, you can ensure that the necessary headers are sent with your GraphQL requests.

**Comment by jengle-dev** [comment]
I asked ChatGPT if it can be combined with the mongoose connection and it gave this: the error still ocurred but maybe the mongoose link isn't working. I don't know. 

Yes, you can integrate the code for setting up Apollo Client with a Mongoose connection in the same JavaScript file. Here's an example of how you can combine both the Apollo Client setup and Mongoose connection in a single file:

```javascript
// Import necessary dependencies
import { ApolloClient, ApolloProvider, InMemoryCache, createHttpLink } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';
import mongoose from 'mongoose';

// Mongoose connection setup
mongoose.connect('mongodb://localhost:27017/mydatabase', {
  useNewUrlParser: true,
  useUnifiedTopology: true,
});
const db = mongoose.connection;
db.on('error', console.error.bind(console, 'MongoDB connection error:'));
db.once('open', () => {
  console.log('MongoDB connected');
});

// Create an HTTP link for Apollo Client
const httpLink = createHttpLink({
  uri: '/graphql',
});

// Create an auth link to add custom headers
const authLink = setContext((_, { headers }) => {
  const customHeaders = {
    'Authorization': 'Bearer YOUR_TOKEN',
    'Custom-Header': 'Custom Value',
  };

  return {
    headers: {
      ...headers,
      ...customHeaders,
    },
  };
});

// Create an Apollo Client instance
const client = new ApolloClient({
  link: authLink.concat(httpLink),
  cache: new InMemoryCache(),
});

// Export the Apollo Client instance and Mongoose connection
export { client, db };
```

In this example, the code sets up both the Mongoose connection and Apollo Client in the same file. You can adjust the Mongoose connection URL (`mongodb://localhost:27017/mydatabase`) to match your specific MongoDB configuration.

By exporting the `client` and `db` variables, you can import and use them in other files where you need to interact with your MongoDB database or make GraphQL requests using Apollo Client.

Remember to install the necessary dependencies (`@apollo/client`, `mongoose`, etc.) and import them at the top of your JavaScript file.

Please note that this is just a simplified example, and you might need to adapt it to your specific project structure and requirements.

**Comment by juliaguas17** [comment]
I added a couple things...

1. added bodyParser const and added line 11 app.use(bodyParser.json(()) - Express body-parser is an npm module used to process data sent in an HTTP request body. It provides four express middleware for parsing JSON, Text, URL-encoded, and raw data sets over an HTTP request body. Before the target controller receives an incoming request, these middleware routines handle it ---> I think this is the main reason we were having the Request Header error is because the controller was trying to handle too many incoming requests
2. Changed "proxy" to localhost:3001 in client-side package.json
3. Line 33 in server file was missing a path for graphql. Originally was just server.applyMiddleware {{ app }}. I changed this to server.applyMiddleware({ app, path: '/graphql' }).
4. created separate constants for typeDefs and resolvers instead of just importing the schema folder (not sure if this made much of a difference but I think it solved the issue where it was looking for a "models" folder)
